File: backend/app/api/routes/executions.py
"""
Execution management API routes (single-user, no auth, no database).
"""

import asyncio
import json
import math
import time
from typing import Optional

from fastapi import APIRouter, HTTPException, Query, status, Depends, WebSocket, WebSocketDisconnect

from app.schemas.execution import (
    ExecutionCreate,
    ExecutionResponse,
    ExecutionListResponse,
    ExecutionControl,
    ExecutionMessageResponse,
)
from app.schemas.common import PaginatedResponse, SuccessResponse
from app.services.execution_service import ExecutionService
from app.store import get_store, LOCAL_USER_ID, Store


router = APIRouter()


@router.get("", response_model=PaginatedResponse[ExecutionListResponse])
async def list_executions(
    store: Store = Depends(get_store),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    team_id: Optional[str] = None,
    status_filter: Optional[str] = None,
):
    service = ExecutionService(store)
    executions, total = service.list(
        user_id=LOCAL_USER_ID,
        page=page,
        page_size=page_size,
        team_id=team_id,
        status_filter=status_filter,
    )

    items = [
        ExecutionListResponse(
            id=e["id"],
            team_id=e.get("team_id"),
            title=e.get("title"),
            status=e.get("status"),
            current_round=e.get("current_round", 0),
            tokens_used=e.get("tokens_used", 0),
            cost=e.get("cost", 0.0),
            started_at=e.get("started_at"),
            completed_at=e.get("completed_at"),
            created_at=e["created_at"],
        )
        for e in executions
    ]

    return PaginatedResponse(
        items=items,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=math.ceil(total / page_size) if total > 0 else 0,
    )


@router.post("", response_model=ExecutionResponse, status_code=status.HTTP_201_CREATED)
async def create_execution(
    execution_data: ExecutionCreate,
    store: Store = Depends(get_store),
):
    service = ExecutionService(store)
    try:
        execution = service.create(LOCAL_USER_ID, execution_data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    return ExecutionResponse.model_validate(execution)


@router.get("/{execution_id}", response_model=ExecutionResponse)
async def get_execution(
    execution_id: str,
    store: Store = Depends(get_store),
):
    service = ExecutionService(store)
    execution = service.get(execution_id, LOCAL_USER_ID)
    if not execution:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Execution {execution_id} not found")

    messages = service.get_messages(execution_id, limit=50)
    response = ExecutionResponse.model_validate(execution)
    response.recent_messages = [ExecutionMessageResponse.model_validate(m) for m in messages]
    return response


@router.get("/{execution_id}/messages", response_model=list[ExecutionMessageResponse])
async def get_execution_messages(
    execution_id: str,
    store: Store = Depends(get_store),
    limit: int = Query(100, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    round_filter: Optional[int] = None,
    agent_id: Optional[str] = None,
):
    service = ExecutionService(store)
    execution = service.get(execution_id, LOCAL_USER_ID)
    if not execution:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Execution {execution_id} not found")

    messages = service.get_messages(
        execution_id,
        limit=limit,
        offset=offset,
        round_filter=round_filter,
        agent_id=agent_id,
    )
    return [ExecutionMessageResponse.model_validate(m) for m in messages]


@router.post("/{execution_id}/control", response_model=SuccessResponse)
async def control_execution(
    execution_id: str,
    control: ExecutionControl,
    store: Store = Depends(get_store),
):
    service = ExecutionService(store)
    success = service.control(execution_id, LOCAL_USER_ID, control.action, control.params)
    if not success:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail="Invalid action or execution state")
    return SuccessResponse(message=f"Action '{control.action}' executed successfully")

@router.websocket("/{execution_id}/ws")
async def execution_ws(
    websocket: WebSocket,
    execution_id: str,
    store: Store = Depends(get_store),
):
    await websocket.accept()
    service = ExecutionService(store)
    execution = service.get(execution_id, LOCAL_USER_ID)
    if not execution:
        await websocket.send_text(
            json.dumps(
                {"event_type": "error", "data": {"message": f"Execution {execution_id} not found"}},
                ensure_ascii=False,
            )
        )
        await websocket.close()
        return

    async def send_event(event):
        await websocket.send_text(
            json.dumps(
                {
                    "event_type": event.event_type,
                    "data": event.data,
                    "agent_id": event.agent_id,
                    "sequence": event.sequence,
                },
                ensure_ascii=False,
            )
        )

    last_seen = time.monotonic()
    heartbeat_timeout = 60.0

    await websocket.send_text(
        json.dumps({"event_type": "connected", "data": {"execution_id": execution_id}}, ensure_ascii=False)
    )

    if execution.get("status") == "pending":
        async for event in service.start(execution_id, LOCAL_USER_ID):
            await send_event(event)
    else:
        await websocket.send_text(
            json.dumps({"event_type": "status", "data": {"status": execution.get("status")}}, ensure_ascii=False)
        )

    try:
        while True:
            try:
                raw = await asyncio.wait_for(websocket.receive_text(), timeout=heartbeat_timeout)
            except asyncio.TimeoutError:
                if time.monotonic() - last_seen >= heartbeat_timeout:
                    await websocket.close()
                    break
                continue

            last_seen = time.monotonic()
            try:
                payload = json.loads(raw)
            except json.JSONDecodeError:
                await websocket.send_text(
                    json.dumps({"event_type": "error", "data": {"message": "Invalid JSON payload"}}, ensure_ascii=False)
                )
                continue

            msg_type = payload.get("type")
            if msg_type == "followup":
                input_text = str(payload.get("input") or "").strip()
                if not input_text:
                    await websocket.send_text(
                        json.dumps(
                            {"event_type": "error", "data": {"message": "Follow-up input required"}},
                            ensure_ascii=False,
                        )
                    )
                    continue
                target_agent_id = payload.get("target_agent_id")
                async for event in service.followup(execution_id, LOCAL_USER_ID, input_text, target_agent_id):
                    await send_event(event)
                await websocket.send_text(
                    json.dumps({"event_type": "status", "data": {"status": "completed"}}, ensure_ascii=False)
                )
            elif msg_type == "start":
                execution = service.get(execution_id, LOCAL_USER_ID)
                if execution and execution.get("status") == "pending":
                    async for event in service.start(execution_id, LOCAL_USER_ID):
                        await send_event(event)
            elif msg_type == "control":
                action = payload.get("action")
                params = payload.get("params") if isinstance(payload.get("params"), dict) else None
                if not action or not service.control(execution_id, LOCAL_USER_ID, action, params):
                    await websocket.send_text(
                        json.dumps(
                            {"event_type": "error", "data": {"message": "Invalid control action"}},
                            ensure_ascii=False,
                        )
                    )
            elif msg_type in {"ping", "pong"}:
                await websocket.send_text(json.dumps({"event_type": "pong", "data": {}}, ensure_ascii=False))
            else:
                await websocket.send_text(
                    json.dumps({"event_type": "error", "data": {"message": "Unknown message type"}}, ensure_ascii=False)
                )
    except WebSocketDisconnect:
        return


@router.delete("/{execution_id}", response_model=SuccessResponse)
async def delete_execution(
    execution_id: str,
    store: Store = Depends(get_store),
):
    service = ExecutionService(store)
    success = service.delete(execution_id, LOCAL_USER_ID)
    if not success:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=f"Execution {execution_id} not found")
    return SuccessResponse(message="Execution deleted successfully")

File: backend/src/commands/executions.rs
use chrono::Utc;
use serde::Serialize;
use serde_json::Value;
use tauri::{Emitter, State, Window};
use uuid::Uuid;

use crate::agents::instance::AgentInstance;
use crate::error::AppError;
use crate::llm::factory::{provider_from_runtime_config, resolve_runtime_config_for_agent};
use crate::models::common::{PaginatedResponse, SuccessResponse};
use crate::models::execution::{ExecutionCreate, ExecutionListItem, ExecutionMessage, ExecutionRecord, ExecutionResponse};
use crate::models::team::Team;
use crate::orchestration::debate::run_debate;
use crate::orchestration::pipeline::run_pipeline;
use crate::orchestration::roundtable::run_roundtable;
use crate::orchestration::state::OrchestrationState;
use crate::state::AppState;

const LOCAL_USER_ID: &str = "local";
const EVENT_NAME: &str = "execution-event";

#[derive(Debug, Clone, Serialize)]
struct ExecutionEventPayload {
    execution_id: String,
    event_type: String,
    data: Value,
    agent_id: Option<String>,
    sequence: u64,
}

#[tauri::command]
pub fn list_executions(
    state: State<AppState>,
    page: Option<usize>,
    page_size: Option<usize>,
    team_id: Option<String>,
    status_filter: Option<String>,
) -> Result<PaginatedResponse<ExecutionListItem>, AppError> {
    let page = page.unwrap_or(1).max(1);
    let page_size = page_size.unwrap_or(20).clamp(1, 100);

    let mut executions = state.store.executions_list()?;
    executions = executions
        .into_iter()
        .filter(|e| e.user_id == LOCAL_USER_ID)
        .collect();

    if let Some(team_id) = team_id {
        executions = executions.into_iter().filter(|e| e.team_id == team_id).collect();
    }
    if let Some(status_filter) = status_filter {
        executions = executions
            .into_iter()
            .filter(|e| e.status == status_filter)
            .collect();
    }

    executions.sort_by(|a, b| b.created_at.cmp(&a.created_at));

    let total = executions.len();
    let total_pages = if total == 0 {
        0
    } else {
        (total + page_size - 1) / page_size
    };
    let start = (page - 1) * page_size;
    let end = (start + page_size).min(total);
    let slice = if start >= total { &[] } else { &executions[start..end] };

    let items = slice
        .iter()
        .map(|e| ExecutionListItem {
            id: e.id.clone(),
            team_id: e.team_id.clone(),
            title: e.title.clone(),
            status: e.status.clone(),
            current_round: e.current_round,
            tokens_used: e.tokens_used,
            cost: e.cost,
            started_at: e.started_at,
            completed_at: e.completed_at,
            created_at: e.created_at,
        })
        .collect();

    Ok(PaginatedResponse {
        items,
        total,
        page,
        page_size,
        total_pages,
    })
}

#[tauri::command]
pub fn get_execution(state: State<AppState>, id: String) -> Result<ExecutionResponse, AppError> {
    let record = state
        .store
        .executions_get(&id)?
        .ok_or_else(|| AppError::Message(format!("Execution {id} not found")))?;
    let mut messages = state.store.execution_messages_list(&id)?;
    messages.sort_by_key(|m| m.sequence);
    let total = messages.len();
    let start = total.saturating_sub(50);
    let recent = messages[start..].to_vec();
    Ok(ExecutionResponse::from_record(record, recent))
}

#[tauri::command]
pub fn create_execution(state: State<AppState>, execution: ExecutionCreate) -> Result<ExecutionResponse, AppError> {
    let now = Utc::now();
    let record = ExecutionRecord {
        id: Uuid::new_v4().to_string(),
        user_id: LOCAL_USER_ID.to_string(),
        team_id: execution.team_id,
        title: execution.title,
        initial_input: execution.input,
        llm: execution.llm,
        status: "pending".to_string(),
        current_stage: None,
        current_round: 0,
        shared_state: serde_json::json!({}),
        agent_states: serde_json::json!({}),
        final_output: None,
        structured_output: None,
        tokens_used: 0,
        tokens_budget: execution.budget.max_tokens,
        cost: 0.0,
        cost_budget: execution.budget.max_cost,
        started_at: None,
        completed_at: None,
        error_message: None,
        retry_count: 0,
        workspace_path: execution.workspace_path,
        created_at: now,
        updated_at: now,
    };
    state.store.executions_upsert(&record)?;
    Ok(ExecutionResponse::from_record(record, Vec::new()))
}

#[tauri::command]
pub fn delete_execution(state: State<AppState>, id: String) -> Result<SuccessResponse, AppError> {
    state.store.executions_delete(&id)?;
    Ok(SuccessResponse {
        success: true,
        message: "Execution deleted successfully".to_string(),
    })
}

#[tauri::command]
pub fn control_execution(
    state: State<AppState>,
    id: String,
    action: String,
    params: Option<Value>,
) -> Result<SuccessResponse, AppError> {
    let mut execution = state
        .store
        .executions_get(&id)?
        .ok_or_else(|| AppError::Message(format!("Execution {id} not found")))?;

    let params = params.unwrap_or_else(|| serde_json::json!({}));
    let status = execution.status.clone();

    if action == "pause" && status == "running" {
        execution.status = "paused".to_string();
    } else if action == "resume" && status == "paused" {
        execution.status = "running".to_string();
    } else if action == "stop" && (status == "running" || status == "paused") {
        execution.status = "completed".to_string();
        execution.completed_at = Some(Utc::now());
    } else if action == "extend_budget" {
        let add_tokens = params.get("tokens").and_then(|v| v.as_u64()).unwrap_or(50_000) as u32;
        let add_cost = params.get("cost").and_then(|v| v.as_f64()).unwrap_or(5.0);
        execution.tokens_budget = execution.tokens_budget.saturating_add(add_tokens);
        execution.cost_budget += add_cost;
    } else {
        return Err(AppError::Message("Invalid action or execution state".to_string()));
    }

    execution.updated_at = Utc::now();
    state.store.executions_upsert(&execution)?;

    Ok(SuccessResponse {
        success: true,
        message: format!("Action '{action}' executed successfully"),
    })
}

#[tauri::command]
pub fn set_execution_workspace(
    state: State<AppState>,
    id: String,
    workspace_path: Option<String>,
) -> Result<ExecutionResponse, AppError> {
    let mut execution = state
        .store
        .executions_get(&id)?
        .ok_or_else(|| AppError::Message(format!("Execution {id} not found")))?;
    execution.workspace_path = workspace_path;
    execution.updated_at = Utc::now();
    state.store.executions_upsert(&execution)?;
    Ok(ExecutionResponse::from_record(execution, Vec::new()))
}

#[tauri::command]
pub fn start_execution(window: Window, state: State<AppState>, execution_id: String) -> Result<(), AppError> {
    let store = state.store.clone();
    let window = window.clone();

    tauri::async_runtime::spawn(async move {
        if let Err(err) = run_execution(window.clone(), store.clone(), execution_id.clone(), None, None).await {
            let message = err.to_string();
            if let Ok(Some(mut execution)) = store.executions_get(&execution_id) {
                execution.status = "failed".to_string();
                execution.error_message = Some(message.clone());
                execution.updated_at = Utc::now();
                let _ = store.executions_upsert(&execution);
            }
            let mut seq = 0;
            emit_event(&window, &execution_id, "error", serde_json::json!({ "message": message }), None, &mut seq);
        }
    });

    Ok(())
}

#[tauri::command]
pub fn followup_execution(
    window: Window,
    state: State<AppState>,
    execution_id: String,
    input: String,
    target_agent_id: Option<String>,
) -> Result<(), AppError> {
    let store = state.store.clone();
    let window = window.clone();

    tauri::async_runtime::spawn(async move {
        if let Err(err) =
            run_execution(window.clone(), store.clone(), execution_id.clone(), Some(input), target_agent_id).await
        {
            let message = err.to_string();
            if let Ok(Some(mut execution)) = store.executions_get(&execution_id) {
                execution.status = "failed".to_string();
                execution.error_message = Some(message.clone());
                execution.updated_at = Utc::now();
                let _ = store.executions_upsert(&execution);
            }
            let mut seq = 0;
            emit_event(&window, &execution_id, "error", serde_json::json!({ "message": message }), None, &mut seq);
        }
    });

    Ok(())
}

async fn run_execution(
    window: Window,
    store: std::sync::Arc<crate::store::sqlite::SqliteStore>,
    execution_id: String,
    followup_input: Option<String>,
    target_agent_id: Option<String>,
) -> Result<(), AppError> {
    let mut event_seq: u64 = 0;

    let Some(mut execution) = store.executions_get(&execution_id)? else {
        emit_event(
            &window,
            &execution_id,
            "error",
            serde_json::json!({"message": "Execution not found"}),
            None,
            &mut event_seq,
        );
        return Ok(());
    };

    if followup_input.is_none() && execution.status != "pending" {
        // Allow users to reconnect / view history without re-starting the execution.
        return Ok(());
    }

    if let Some(input) = &followup_input {
        if execution.status != "paused" && execution.status != "completed" {
            emit_event(
                &window,
                &execution_id,
                "error",
                serde_json::json!({"message": "Invalid execution state for follow-up"}),
                None,
                &mut event_seq,
            );
            return Ok(());
        }
        // Continue with follow-up.
        execution.status = "running".to_string();
        execution.updated_at = Utc::now();
        store.executions_upsert(&execution)?;
        emit_event(
            &window,
            &execution_id,
            "status",
            serde_json::json!({"status": "running"}),
            None,
            &mut event_seq,
        );

        run_round(window, store, execution, input.clone(), target_agent_id, &mut event_seq).await?;
        return Ok(());
    }

    // Start execution
    let initial = execution.initial_input.trim().to_string();
    if initial.is_empty() {
        execution.status = "paused".to_string();
        execution.updated_at = Utc::now();
        store.executions_upsert(&execution)?;
        emit_event(
            &window,
            &execution_id,
            "status",
            serde_json::json!({
                "message": "等待输入讨论主题（请输入内容后发送追问以继续）",
                "phase": "awaiting_user_input"
            }),
            None,
            &mut event_seq,
        );
        return Ok(());
    }

    execution.status = "running".to_string();
    execution.started_at = Some(Utc::now());
    execution.updated_at = Utc::now();
    store.executions_upsert(&execution)?;

    emit_event(
        &window,
        &execution_id,
        "status",
        serde_json::json!({"status": "running"}),
        None,
        &mut event_seq,
    );

    run_round(window, store, execution, initial, None, &mut event_seq).await?;
    Ok(())
}

async fn run_round(
    window: Window,
    store: std::sync::Arc<crate::store::sqlite::SqliteStore>,
    mut execution: ExecutionRecord,
    topic: String,
    target_agent_id: Option<String>,
    event_seq: &mut u64,
) -> Result<(), AppError> {
    let execution_id = execution.id.clone();
    let team = store
        .teams_get(&execution.team_id)?
        .ok_or_else(|| AppError::Message("Team not found".to_string()))?;

    let Some(llm) = execution.llm.clone() else {
        emit_event(
            &window,
            &execution_id,
            "error",
            serde_json::json!({"message": "No LLM configured. Please set it in the UI (API配置) and start a new execution."}),
            None,
            event_seq,
        );
        execution.status = "failed".to_string();
        execution.error_message = Some("No LLM configured".to_string());
        execution.updated_at = Utc::now();
        store.executions_upsert(&execution)?;
        return Ok(());
    };

    let mut state: OrchestrationState = serde_json::from_value(execution.shared_state.clone()).unwrap_or_default();
    if state.topic.trim().is_empty() {
        state.topic = topic.clone();
    }
    state.start_new_round();
    state.topic = topic.clone();
    let round_num = state.round;

    let agents = build_agent_instances(&store, &team, &llm, target_agent_id.as_deref()).await?;

    let mut msg_seq = store.execution_messages_next_sequence(&execution_id)?;

    // persist + emit user message first
    let now = Utc::now();
    let user_message = ExecutionMessage {
        id: Uuid::new_v4().to_string(),
        sequence: msg_seq,
        round: state.round,
        phase: "user".to_string(),
        sender_type: "user".to_string(),
        sender_id: None,
        sender_name: Some("you".to_string()),
        content: topic.clone(),
        content_type: "text".to_string(),
        responding_to: None,
        target_agent_id: None,
        confidence: None,
        wants_to_continue: true,
        input_tokens: 0,
        output_tokens: 0,
        metadata: serde_json::json!({}),
        created_at: now,
        updated_at: now,
    };
    store.execution_messages_upsert(&execution_id, &user_message)?;
    emit_event(
        &window,
        &execution_id,
        "user",
        serde_json::json!({
            "content": topic,
            "phase": "user",
            "round": state.round,
            "message_id": user_message.id,
            "message_sequence": user_message.sequence
        }),
        None,
        event_seq,
    );
    msg_seq += 1;

    let mut emit = |event_type: &str, mut data: Value, agent_id: Option<String>| -> Result<(), AppError> {
        if event_type == "opinion" {
            let agent_name = data.get("agent_name").and_then(|v| v.as_str()).unwrap_or("agent");
            let content = data.get("content").and_then(|v| v.as_str()).unwrap_or("");
            let round = data.get("round").and_then(|v| v.as_i64()).unwrap_or(round_num as i64) as i32;
            let phase = data.get("phase").and_then(|v| v.as_str()).unwrap_or("opinion").to_string();
            let now = Utc::now();
            let message = ExecutionMessage {
                id: Uuid::new_v4().to_string(),
                sequence: msg_seq,
                round,
                phase: phase.clone(),
                sender_type: if agent_id.is_some() { "agent".to_string() } else { "system".to_string() },
                sender_id: agent_id.clone(),
                sender_name: Some(agent_name.to_string()),
                content: content.to_string(),
                content_type: "text".to_string(),
                responding_to: data.get("responding_to").and_then(|v| v.as_str()).map(|s| s.to_string()),
                target_agent_id: data.get("target_agent_id").and_then(|v| v.as_str()).map(|s| s.to_string()),
                confidence: data.get("confidence").and_then(|v| v.as_f64()),
                wants_to_continue: data.get("wants_to_continue").and_then(|v| v.as_bool()).unwrap_or(true),
                input_tokens: data.get("input_tokens").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
                output_tokens: data.get("output_tokens").and_then(|v| v.as_i64()).unwrap_or(0) as i32,
                metadata: data.get("metadata").cloned().unwrap_or_else(|| serde_json::json!({})),
                created_at: now,
                updated_at: now,
            };
            store.execution_messages_upsert(&execution_id, &message)?;
            if let Some(obj) = data.as_object_mut() {
                obj.insert("message_id".to_string(), Value::String(message.id.clone()));
                obj.insert("message_sequence".to_string(), serde_json::json!(message.sequence));
            }
            msg_seq += 1;
        }

        emit_event(&window, &execution_id, event_type, data, agent_id, event_seq);
        Ok(())
    };

    // Choose orchestrator
    match team.collaboration_mode.as_str() {
        "pipeline" => {
            state.phase = crate::orchestration::state::OrchestrationPhase::Sequential;
            let _ = run_pipeline(agents, &mut state, &mut emit).await?;
        }
        "debate" => {
            let _ = run_debate(agents, &mut state, &mut emit, 3).await?;
        }
        _ => {
            let _ = run_roundtable(agents, &mut state, &mut emit, true).await?;
        }
    }

    // Save execution state
    execution.status = "completed".to_string();
    execution.completed_at = Some(Utc::now());
    execution.current_round = state.round;
    execution.tokens_used = state.tokens_used;
    execution.shared_state = serde_json::to_value(&state).unwrap_or_else(|_| serde_json::json!({}));
    execution.updated_at = Utc::now();
    store.executions_upsert(&execution)?;

    emit_event(
        &window,
        &execution_id,
        "status",
        serde_json::json!({"status": "completed"}),
        None,
        event_seq,
    );

    Ok(())
}

async fn build_agent_instances(
    store: &crate::store::sqlite::SqliteStore,
    team: &Team,
    llm: &crate::models::llm::ExecutionLLMConfig,
    target_agent_id: Option<&str>,
) -> Result<Vec<AgentInstance>, AppError> {
    let mut members = team.members.clone();
    members.sort_by_key(|m| m.position);
    let agent_ids = members
        .iter()
        .filter(|m| m.is_active)
        .map(|m| m.agent_id.clone())
        .collect::<Vec<_>>();

    let mut instances = Vec::new();
    for agent_id in agent_ids {
        if let Some(target) = target_agent_id {
            if agent_id != target {
                continue;
            }
        }
        let Some(agent) = store.agents_get(&agent_id)? else {
            continue;
        };

        let cfg = resolve_runtime_config_for_agent(agent.model_id.as_deref(), llm)?;
        let provider = provider_from_runtime_config(&cfg)?;
        instances.push(AgentInstance::from_agent(&agent, provider));
    }

    if instances.is_empty() {
        return Err(AppError::Message("No agents in team".to_string()));
    }

    Ok(instances)
}

fn emit_event(
    window: &Window,
    execution_id: &str,
    event_type: &str,
    data: Value,
    agent_id: Option<String>,
    seq: &mut u64,
) {
    *seq += 1;
    let payload = ExecutionEventPayload {
        execution_id: execution_id.to_string(),
        event_type: event_type.to_string(),
        data,
        agent_id,
        sequence: *seq,
    };
    let _ = window.emit(EVENT_NAME, payload);
}
